bin/pip_find_builddeps.py

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python3
+import argparse
+import contextlib
+import datetime
+import re
+import shutil
+import subprocess
+import sys
+import tempfile
+from pathlib import Path
+
+SCRIPT_NAME = Path(sys.argv[0]).name
+
+
+def _pip_download(requirements_files, output_file, tmpdir, no_cache):
+    """Run pip download, write output to file."""
+    cmd = ["pip", "download", "-d", tmpdir, "--no-binary", ":all:", "--verbose"]
+    if no_cache:
+        cmd.append("--no-cache-dir")
+    for file in requirements_files:
+        cmd.append("-r")
+        cmd.append(file)
+
+    with open(output_file, "w") as outfile:
+        subprocess.run(cmd, stdout=outfile, stderr=outfile, check=True)
+
+
+def _filter_builddeps(pip_download_output_file):
+    """Find builddeps in output of pip download."""
+    # Leading whitespace => package is a build dependency
+    # (because all recursive runtime dependencies were present in input files)
+    builddep_re = re.compile(r"^\s+Collecting (\S+)")
+
+    with open(pip_download_output_file) as f:
+        matches = (builddep_re.match(line) for line in f)
+        builddeps = set(match.group(1) for match in matches if match)
+
+    return sorted(builddeps)
+
+
+@contextlib.contextmanager
+def _remove_on_success(tmpdir):
+    """Remove tmpdir only if no exception is raised during context."""
+    try:
+        yield
+    except:  # noqa: E722
+        # Re-raise the exception but keep the tmpdir
+        raise
```

Comment:
I am wondering if it would be better to print the error message to STDERR end exit with a code > 0 here for so it is easier for a human user to parse the output.

Then, for the most common error here (which I suppose to be dependencies not being fetched), the user would see just:

`ERROR: pip download failed, see /tmp/pip_find_builddeps.py-j0o4ghx0/pip-download-output.txt for more info`

Instead of the stack trace.

A debug mode could output the full trace though.

Reply:
Good idea, I instead added a custom exception type and I'm suppressing it for main() execution. This way, expected errors will look nice and unexpected errors will still produce a full traceback